DevTools Week: Doc page for innerloop versus outerloop guide (#80)
devfile/devfile-web: libs/docs/src/docs/2.2.0/innerloop-vs-outerloop.md

---
title: Innerloop versus outerloop
description: Innerloop versus outerloop
---

In a devfile spec, there are two scopes of deployment: innerloop and outerloop. Having these scopes is essential for a full development experience as well as ensuring proper integration of the full scope of development tools for Kubernetes and OpenShift projects.

## Prerequisites
- [Adding components](./adding-components)
- [Adding commands](./adding-commands)

## What is innerloop?

**Innerloop** deployments are actions the developer makes within their development environment, such as running tests, debugging, and local deployments before checking into the target source repository.

### Procedure

1. Add the runtime container component for running your project with the `schemaVersion` and `metadata` definitions
    - This example will use `nodejs-18` UBI image

    ```yaml
    schemaVersion: 2.2.0
    metadata:
      name: nodejs
    components:
      - name: runtime
        container:
          image: registry.access.redhat.com/ubi8/nodejs-18:latest
          args: ['tail', '-f', '/dev/null']
          memoryLimit: 1024Mi
          mountSources: true
    ```

> 2. If your project needs ports to forwarded like in this example, define the endpoints required for your project

Reviewer: ```suggestion
2. If your project needs ports to be forwarded like in this example, define the endpoints required for your project
```

    - ***Note**: If you are using odov3, you need to create an additional endpoint for the debugging port. See [Odo Spec Support](#odo-spec-support) for more information*
    - The environment variable definition `DEBUG_PORT` is to set the debug port of the node application

    ```yaml
    components:
      - name: runtime
        container:
          image: registry.access.redhat.com/ubi8/nodejs-18:latest
          args: ['tail', '-f', '/dev/null']
          memoryLimit: 1024Mi
          mountSources: true
          env:
            - name: DEBUG_PORT
              value: '5858'
          endpoints:
            - name: http-node
              targetPort: 3000
            - exposure: none
              name: debug
              targetPort: 5858
    ```

3. Now the runtime component has been defined, we need a way of running our project
    1. Add a command to `npm install` our project packages and mark it as our projects build command
    2. Add another command to run the main process or the server process `npm start` in this example and mark it as our projects run command

>     3. With these command you should now be able to run a development deployment of your project, for example `odo dev`

Reviewer: ```suggestion
    3. With these commands you should now be able to run a development deployment of your project, for example `odo dev`
```

          group:
            kind: run
            isDefault: true
    ```

4. For testing and debugging we will also need to define commands for each with the shell commands used to debug and test your project
    - The `kind` of commands will be `debug` and `test` for running your debugging and testing respectively

    ```yaml
    commands:
      - id: install
        exec:
          component: runtime
          commandLine: npm install
          workingDir: ${PROJECT_SOURCE}
          group:
            kind: build
            isDefault: true
      - id: run
        exec:
          component: runtime
          commandLine: npm start
          workingDir: ${PROJECT_SOURCE}
          group:
            kind: run
            isDefault: true
      - id: debug
        exec:
          component: runtime
          commandLine: npm run debug
          workingDir: ${PROJECT_SOURCE}
          group:
            kind: debug
            isDefault: true
      - id: test
        exec:
          component: runtime
          commandLine: npm test
          workingDir: ${PROJECT_SOURCE}
          group:
            kind: test
            isDefault: true
    ```

> ```yaml {% title="Final innerloop devfile" filename="devfile.yaml" %}

Reviewer: add a header above the final devfile snippet

I wonder if it would be better to show the full devfile first and then the procedure

Author: @mike-hoang

> add a header above the final devfile snippet

The `title` field adds a header above the code snippet, I could change this to a markdown header if you feel it can be shown a bit better.

> I wonder if it would be better to show the full devfile first and then the procedure

From my own perspective I disagree with this as I feel it would be overwhelming for a reader to see the full source block at the beginning, *however* I think this opens a pretty good discussion to have about what would be a better format for more readable documentation. Currently, the doc pages I've worked with are setup with the full source at the end as what is seen here.

Reviewer: I'm ok with keeping the format as is, just wanted to bring it up as I've seen it done both ways before

schemaVersion: 2.2.0
metadata:
  name: nodejs
components:
  - name: runtime
    container:
      image: registry.access.redhat.com/ubi8/nodejs-18:latest
      args: ['tail', '-f', '/dev/null']
      memoryLimit: 1024Mi
      mountSources: true
      env:
        - name: DEBUG_PORT
          value: '5858'
      endpoints:
        - name: http-node
          targetPort: 3000
        - exposure: none
          name: debug
          targetPort: 5858
commands:
  - id: install
    exec:
      component: runtime
      commandLine: npm install
      workingDir: ${PROJECT_SOURCE}
      group:
        kind: build
        isDefault: true
  - id: run
    exec:
      component: runtime
      commandLine: npm start
      workingDir: ${PROJECT_SOURCE}
      group:
        kind: run
        isDefault: true
  - id: debug
    exec:
      component: runtime
      commandLine: npm run debug
      workingDir: ${PROJECT_SOURCE}
      group:
        kind: debug
        isDefault: true
  - id: test
    exec:
      component: runtime
      commandLine: npm test
      workingDir: ${PROJECT_SOURCE}
      group:
        kind: test
        isDefault: true
```

> The component and commands here allows the developer to build, run, debug, and test their project within a local cluster using a devfile supported development tool (e.g. `odo`).

Reviewer: ```suggestion
The component and commands here allow the developer to build, run, debug, and test their project within a local cluster using a devfile supported development tool (e.g. `odo`).
```

    1. Provide the file path to your openshift template file with `uri`

    ```yaml
    components:
      - name: outerloop-build
        image:
          imageName: landingpage-image:latest
            dockerfile:
              uri: docker/Dockerfile
              buildContext: .
              rootRequired: false
      - name: outerloop-deploy
        openshift:
          uri: landingpage-template.yaml
    ```

4. As with the image component, we will need to create a command for the openshift component

    ```yaml
    commands:
      - id: build-image
        apply:
          component: outerloop-build
      - id: deploy-openshift
        apply:
          component: outerloop-deploy
    ```

5. To complete the deployment process, combine the two commands into a complete deploy command using composite

    ```yaml
    commands:
      - id: build-image
        apply:
          component: outerloop-build
      - id: deploy-openshift
        apply:
          component: outerloop-deploy
      - id: deploy
        composite:
          commands:
            - build-image
            - deploy-openshift
          group:
            kind: deploy
            isDefault: true
    ```

> ```yaml {% title="Final outerloop devfile" filename="devfile.yaml" %}

Reviewer: add header here too

Reviewer: Sounds good, from a quick glance at the raw file I didn't realize that the title field adds a header

## Support list of developer tools

Though the devfile spec does support both [innerloop](#what-is-innerloop) and [outerloop](#what-is-outerloop) deployments, not all developer tools will support these deployments in the same way. This section will list the devfile supported development tools and what deployment scopes they support.

### Developer Tools

| Tool | Innerloop Support | Outerloop Support |
|------|-------------------|-------------------|
| Odo v2 | X | X |
| Odo v3 | X | X |
| Eclipse Che | X | |
| Amazon CodeCatalyst | X | |
| JetBrains Space Cloud Dev | X | X (*image build only*) |
| Red Hat OpenShift Dev Spaces | X | |
| OpenShift Dev Console | | X |
| VSCode OpenShift Toolkit | X | X |
| IntelliJ OpenShift Toolkit | X | X |

## Odo Spec Support

Odo covers the features supported by the devfile 2.2.0 spec with minor requirements for version 3: 
1. In order to debug your project it is required to define a debug port: [https://odo.dev/docs/user-guides/v3-migration-guide#changes-to-the-way-component-debugging-works](https://odo.dev/docs/user-guides/v3-migration-guide#changes-to-the-way-component-debugging-works)

> 2. By default, odov3 sets up an underlying persistent volumes rather than setting up ephemeral volumes, this can toggled back: [https://odo.dev/docs/user-guides/v3-migration-guide#ephemeral-storage](https://odo.dev/docs/user-guides/v3-migration-guide#ephemeral-storage)

Reviewer: ```suggestion
2. By default, odo v3 sets up underlying persistent volumes rather than setting up ephemeral volumes, this can be toggled back: [https://odo.dev/docs/user-guides/v3-migration-guide#ephemeral-storage](https://odo.dev/docs/user-guides/v3-migration-guide#ephemeral-storage)
```
